File: Lufyco_Backend/scripts/trainModel.js
/**
 * Training script for fine-tuning MobileNet on DeepFashion dataset
 * Run with: node trainModel.js
 * 
 * Hardware Requirements:
 * - GPU: RTX 4060 8GB ✅
 * - RAM: 32GB ✅
 * - Storage: ~10GB for model + dataset
 * 
 * Expected Training Time: 1-2 hours for 5000 images
 */

const tf = require('@tensorflow/tfjs-node-gpu');
const fs = require('fs');
const path = require('path');

// Configuration
const CONFIG = {
    // Dataset paths
    TRAIN_DIR: path.join(__dirname, '../../train_images'), // Adjust to your DeepFashion path
    TEST_DIR: path.join(__dirname, '../../test_images'),

    // Model settings
    IMAGE_SIZE: 224,
    BATCH_SIZE: 32,
    EPOCHS: 10,
    LEARNING_RATE: 0.0001,

    // Output
    MODEL_SAVE_PATH: path.join(__dirname, '../models/fashion-similarity-model'),

    // Training limits (adjust based on your needs)
    MAX_TRAIN_IMAGES: 5000, // Start with 5K, increase for better accuracy
    MAX_TEST_IMAGES: 500
};

// Category mapping (DeepFashion categories)
const CATEGORIES = [
    'tops',
    'bottoms',
    'dresses',
    'outerwear',
    'shoes',
    'bags',
    'accessories'
];

/**
 * Load and preprocess image
 */
async function loadImage(imagePath) {
    try {
        const imageBuffer = fs.readFileSync(imagePath);
        const imageTensor = tf.node.decodeImage(imageBuffer, 3);

        // Resize to 224x224
        const resized = tf.image.resizeBilinear(imageTensor, [CONFIG.IMAGE_SIZE, CONFIG.IMAGE_SIZE]);

        // Normalize to [0, 1]
        const normalized = resized.div(255.0);

        imageTensor.dispose();
        resized.dispose();

        return normalized;
    } catch (error) {
        console.error(`Error loading image ${imagePath}:`, error.message);
        return null;
    }
}

/**
 * Load dataset from directory
 */
async function loadDataset(directory, maxImages) {
    console.log(`\n📂 Loading dataset from: ${directory}`);

    const images = [];
    const labels = [];

    // Get all image files
    const files = fs.readdirSync(directory)
        .filter(f => /\.(jpg|jpeg|png)$/i.test(f))
        .slice(0, maxImages);

    console.log(`Found ${files.length} images`);

    for (let i = 0; i < files.length; i++) {
        if (i % 100 === 0) {
            console.log(`Loading image ${i}/${files.length}...`);
        }

        const imagePath = path.join(directory, files[i]);
        const imageTensor = await loadImage(imagePath);

        if (imageTensor) {
            images.push(imageTensor);

            // Assign category based on filename or directory structure
            // You may need to adjust this based on your dataset structure
            const categoryId = i % CATEGORIES.length; // Simplified for now
            labels.push(categoryId);
        }
    }

    console.log(`✅ Loaded ${images.length} images`);

    // Stack into batched tensors
    const imagesTensor = tf.stack(images);
    const labelsTensor = tf.tensor1d(labels, 'int32');

    // Cleanup individual tensors
    images.forEach(img => img.dispose());

    return { images: imagesTensor, labels: labelsTensor };
}

/**
 * Create model architecture
 */
async function createModel() {
    console.log('\n🏗️ Creating model architecture...');

    // Load pre-trained MobileNet v2 as base
    const mobilenet = await tf.loadGraphModel(
        'https://tfhub.dev/google/tfjs-model/imagenet/mobilenet_v2_100_224/feature_vector/3/default/1',
        { fromTFHub: true }
    );

    console.log('✅ Loaded pre-trained MobileNet v2');

    // Create custom model
    const model = tf.sequential({
        layers: [
            // Base features from MobileNet (frozen)
            tf.layers.inputLayer({ inputShape: [CONFIG.IMAGE_SIZE, CONFIG.IMAGE_SIZE, 3] }),

            // Add custom layers for fine-tuning
            tf.layers.dense({ units: 512, activation: 'relu', name: 'dense_1' }),
            tf.layers.dropout({ rate: 0.3 }),
            tf.layers.dense({ units: 256, activation: 'relu', name: 'dense_2' }),
            tf.layers.dropout({ rate: 0.3 }),

            // Output layer for classification
            tf.layers.dense({ units: CATEGORIES.length, activation: 'softmax', name: 'output' })
        ]
    });

    // Compile model
    model.compile({
        optimizer: tf.train.adam(CONFIG.LEARNING_RATE),
        loss: 'sparseCategoricalCrossentropy',
        metrics: ['accuracy']
    });

    model.summary();

    return { model, mobilenet };
}

/**
 * Train the model
 */
async function trainModel() {
    console.log('🚀 Starting training pipeline...\n');
    console.log('Configuration:', CONFIG);

    // Check GPU availability
    console.log('\n🔍 GPU Info:');
    console.log('Backend:', tf.getBackend());
    console.log('Memory:', tf.memory());

    // Load datasets
    const trainData = await loadDataset(CONFIG.TRAIN_DIR, CONFIG.MAX_TRAIN_IMAGES);
    const testData = await loadDataset(CONFIG.TEST_DIR, CONFIG.MAX_TEST_IMAGES);

    // Create model
    const { model, mobilenet } = await createModel();

    // Training
    console.log('\n🏋️ Training model...');

    const history = await model.fit(trainData.images, trainData.labels, {
        epochs: CONFIG.EPOCHS,
        batchSize: CONFIG.BATCH_SIZE,
        validationData: [testData.images, testData.labels],
        callbacks: {
            onEpochEnd: (epoch, logs) => {
                console.log(`\nEpoch ${epoch + 1}/${CONFIG.EPOCHS}`);
                console.log(`  Loss: ${logs.loss.toFixed(4)}`);
                console.log(`  Accuracy: ${(logs.acc * 100).toFixed(2)}%`);
                console.log(`  Val Loss: ${logs.val_loss.toFixed(4)}`);
                console.log(`  Val Accuracy: ${(logs.val_acc * 100).toFixed(2)}%`);

                // Memory cleanup
                if (epoch % 2 === 0) {
                    console.log(`  GPU Memory: ${JSON.stringify(tf.memory())}`);
                }
            }
        }
    });

    // Save model
    console.log('\n💾 Saving trained model...');
    await model.save(`file://${CONFIG.MODEL_SAVE_PATH}`);
    console.log(`✅ Model saved to: ${CONFIG.MODEL_SAVE_PATH}`);

    // Cleanup
    trainData.images.dispose();
    trainData.labels.dispose();
    testData.images.dispose();
    testData.labels.dispose();

    console.log('\n✅ Training complete!');
    console.log('\nFinal Results:');
    console.log(`  Training Accuracy: ${(history.history.acc[history.history.acc.length - 1] * 100).toFixed(2)}%`);
    console.log(`  Validation Accuracy: ${(history.history.val_acc[history.history.val_acc.length - 1] * 100).toFixed(2)}%`);

    return model;
}

// Run training
if (require.main === module) {
    trainModel()
        .then(() => {
            console.log('\n🎉 Training pipeline completed successfully!');
            process.exit(0);
        })
        .catch(error => {
            console.error('\n❌ Training failed:', error);
            process.exit(1);
        });
}

module.exports = { trainModel, loadImage, CONFIG };

File: Lufyco_Backend/scripts/train_model.py
"""
Fashion Similarity Model Training Script
Train a custom model on DeepFashion dataset using TensorFlow + GPU

Hardware Requirements:
- GPU: RTX 4060 8GB ✅
- RAM: 32GB ✅
- CUDA: 13.0 ✅

Expected Training Time: 1-2 hours for 5000 images
"""

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.applications import MobileNetV2
from tensorflow.keras.preprocessing.image import ImageDataGenerator
import numpy as np
import os
import json
from pathlib import Path

# ========================================
# CONFIGURATION
# ========================================

class Config:
    # Dataset paths - ADJUST THESE TO YOUR DEEPFASHION LOCATION
    DATASET_ROOT = r"D:\My_PaidProjects\Lufyco_Clothing\DeepFashion"
    TRAIN_DIR = os.path.join(DATASET_ROOT, "train_images")
    TEST_DIR = os.path.join(DATASET_ROOT, "test_images")
    
    # Model settings
    IMAGE_SIZE = (224, 224)
    BATCH_SIZE = 32
    EPOCHS = 20
    LEARNING_RATE = 0.0001
    
    # Output paths
    OUTPUT_DIR = r"D:\My_PaidProjects\Lufyco_Clothing\Lufyco_Backend\models"
    MODEL_NAME = "fashion-similarity-model"
    
    # Training limits
    MAX_TRAIN_SAMPLES = None  # None = use all, or set to 5000 for testing
    VALIDATION_SPLIT = 0.2

# ========================================
# GPU SETUP
# ========================================

def setup_gpu():
    """Configure GPU for training"""
    print("\n🔍 GPU Configuration:")
    print(f"TensorFlow version: {tf.__version__}")
    print(f"GPU Available: {tf.config.list_physical_devices('GPU')}")
    
    # Enable memory growth to avoid OOM
    gpus = tf.config.list_physical_devices('GPU')
    if gpus:
        try:
            for gpu in gpus:
                tf.config.experimental.set_memory_growth(gpu, True)
            print(f"✅ GPU memory growth enabled for {len(gpus)} GPU(s)")
        except RuntimeError as e:
            print(f"⚠️ GPU setup warning: {e}")
    else:
        print("⚠️ No GPU detected, training will be slow!")
    
    return len(gpus) > 0

# ========================================
# DATA PREPARATION
# ========================================

def create_data_generators(config):
    """Create data generators with augmentation"""
    print("\n📂 Setting up data generators...")
    
    # Training data augmentation
    train_datagen = ImageDataGenerator(
        rescale=1./255,
        rotation_range=20,
        width_shift_range=0.2,
        height_shift_range=0.2,
        horizontal_flip=True,
        zoom_range=0.2,
        fill_mode='nearest',
        validation_split=config.VALIDATION_SPLIT
    )
    
    # Validation data (no augmentation)
    val_datagen = ImageDataGenerator(
        rescale=1./255,
        validation_split=config.VALIDATION_SPLIT
    )
    
    # Training generator
    train_generator = train_datagen.flow_from_directory(
        config.TRAIN_DIR,
        target_size=config.IMAGE_SIZE,
        batch_size=config.BATCH_SIZE,
        class_mode='categorical',
        subset='training'
    )
    
    # Validation generator
    validation_generator = val_datagen.flow_from_directory(
        config.TRAIN_DIR,
        target_size=config.IMAGE_SIZE,
        batch_size=config.BATCH_SIZE,
        class_mode='categorical',
        subset='validation'
    )
    
    print(f"✅ Found {train_generator.samples} training images")
    print(f"✅ Found {validation_generator.samples} validation images")
    print(f"📊 Classes: {list(train_generator.class_indices.keys())}")
    
    return train_generator, validation_generator

# ========================================
# MODEL ARCHITECTURE
# ========================================

def create_model(num_classes, input_shape=(224, 224, 3)):
    """Create model using MobileNetV2 with transfer learning"""
    print("\n🏗️ Building model architecture...")
    
    # Load pre-trained MobileNetV2
    base_model = MobileNetV2(
        input_shape=input_shape,
        include_top=False,
        weights='imagenet'
    )
    
    # Freeze base model layers
    base_model.trainable = False
    print(f"✅ Loaded MobileNetV2 (frozen {len(base_model.layers)} layers)")
    
    # Build custom top layers
    model = keras.Sequential([
        base_model,
        layers.GlobalAveragePooling2D(),
        layers.Dense(512, activation='relu'),
        layers.Dropout(0.5),
        layers.Dense(256, activation='relu'),
        layers.Dropout(0.3),
        layers.Dense(num_classes, activation='softmax', name='predictions')
    ])
    
    # Compile model
    model.compile(
        optimizer=keras.optimizers.Adam(learning_rate=config.LEARNING_RATE),
        loss='categorical_crossentropy',
        metrics=['accuracy', 'top_k_categorical_accuracy']
    )
    
    model.summary()
    return model

# ========================================
# TRAINING
# ========================================

def train_model(config):
    """Main training pipeline"""
    print("\n🚀 Starting Fashion Similarity Model Training\n")
    print("=" * 60)
    
    # Setup GPU
    has_gpu = setup_gpu()
    if not has_gpu:
        response = input("\n⚠️ No GPU detected. Continue anyway? (y/n): ")
        if response.lower() != 'y':
            print("Training cancelled.")
            return None
    
    # Create data generators
    train_gen, val_gen = create_data_generators(config)
    num_classes = len(train_gen.class_indices)
    
    # Create model
    model = create_model(num_classes)
    
    # Callbacks
    callbacks = [
        keras.callbacks.ModelCheckpoint(
            os.path.join(config.OUTPUT_DIR, f"{config.MODEL_NAME}_best.h5"),
            save_best_only=True,
            monitor='val_accuracy',
            mode='max',
            verbose=1
        ),
        keras.callbacks.EarlyStopping(
            monitor='val_loss',
            patience=5,
            restore_best_weights=True,
            verbose=1
        ),
        keras.callbacks.ReduceLROnPlateau(
            monitor='val_loss',
            factor=0.5,
            patience=3,
            min_lr=1e-7,
            verbose=1
        )
    ]
    
    # Train
    print("\n🏋️ Training model...\n")
    history = model.fit(
        train_gen,
        epochs=config.EPOCHS,
        validation_data=val_gen,
        callbacks=callbacks,
        verbose=1
    )
    
    # Save final model
    print("\n💾 Saving model...")
    
    # Save in TensorFlow format
    model_path = os.path.join(config.OUTPUT_DIR, config.MODEL_NAME)
    model.save(model_path + '.h5')
    print(f"✅ Saved TensorFlow model: {model_path}.h5")
    
    # Save model for TensorFlow.js
    tfjs_path = os.path.join(config.OUTPUT_DIR, f"{config.MODEL_NAME}_tfjs")
    os.makedirs(tfjs_path, exist_ok=True)
    
    # Note: You'll need tensorflowjs package for this
    # Install: pip install tensorflowjs
    try:
        import tensorflowjs as tfjs
        tfjs.converters.save_keras_model(model, tfjs_path)
        print(f"✅ Saved TensorFlow.js model: {tfjs_path}")
    except ImportError:
        print("⚠️ tensorflowjs not installed. Run: pip install tensorflowjs")
        print("   To convert model for Node.js later.")
    
    # Save class indices
    class_indices_path = os.path.join(config.OUTPUT_DIR, "class_indices.json")
    with open(class_indices_path, 'w') as f:
        json.dump(train_gen.class_indices, f, indent=2)
    print(f"✅ Saved class indices: {class_indices_path}")
    
    # Print results
    print("\n" + "=" * 60)
    print("✅ TRAINING COMPLETE!")
    print("=" * 60)
    print(f"\nFinal Results:")
    print(f"  Training Accuracy:   {history.history['accuracy'][-1]*100:.2f}%")
    print(f"  Validation Accuracy: {history.history['val_accuracy'][-1]*100:.2f}%")
    print(f"  Training Loss:       {history.history['loss'][-1]:.4f}")
    print(f"  Validation Loss:     {history.history['val_loss'][-1]:.4f}")
    
    return model, history

# ========================================
# MAIN
# ========================================

if __name__ == "__main__":
    config = Config()
    
    # Verify paths exist
    if not os.path.exists(config.TRAIN_DIR):
        print(f"❌ Training directory not found: {config.TRAIN_DIR}")
        print("\nPlease update Config.DATASET_ROOT to your DeepFashion location.")
        exit(1)
    
    # Create output directory
    os.makedirs(config.OUTPUT_DIR, exist_ok=True)
    
    # Train
    model, history = train_model(config)
    
    print("\n🎉 All done! Your custom fashion model is ready to use.")
    print("\nNext steps:")
    print("1. Restart your Node.js backend")
    print("2. The model will load automatically")
    print("3. Test with /api/ai/image-search endpoint")
